source/agora/common/Amount.d

```diff
@@ -169,6 +169,34 @@ public struct Amount
         return true;
     }
 
+    /***************************************************************************
+
+        Divide this `Amount` into `Q` equal parts and returns the remainder
+
+        This function allows to losslessly partition an `Amount`.
+        Usage example include distributing an `Amount` evenly among participant,
+        or computing an absolute majority of 2/3rd.
+
+        Params:
+            value = The value to divide by.
+                    Dividing by `0` is the same as dividing by `1`.
```

Comment:
```suggestion
                    Dividing by `0` is the same as dividing by `1`, and will return Amount(0)
```

Maybe it should be named `split` if it **must** split a value into different parts?

I'm not sure. I've never seen a math function fail to divide by 1, but I'm sure you have a good reason for it. :)

Reply:
Ah it returns the remainder, I'm sorry I misread it!

Reply:
Is there a particular reason why all the operations have side-effects rather than returning a new value?

Reply:
Because if we wanted them to create new values, it would be:
```D
bool div (ref Amount result, uint value) const;
```

Because we don't want to `return` the value (to avoid misuse of the API). And it felt like, if you need a new value, you can just copy it. But I'm open to suggestions.

Reply:
> to avoid misuse of the API

That is what I'm wondering about. What are the examples of misusing it?

My only concern is that these operators do not compose well with range routines like `map` and I can't use them with `const` values, like `const amount = left + right`. In general I prefer functions with no side-effects.

Reply:
```D
bool b;
auto result = myAmount.add( Amount(84), b);
assert(b);
// ...
auto res = result.add(Amount(42), b);
// No check for `b` here
```

In general it's easier to miss checking a parameter than a return value.
Regarding `map`, is it for unittests or production code ? For the former you can just use `Amount(cost_val).mustAdd(another_amount)`.

Reply:
Right, this is good.

I think we really need to propose a `@noignore` attribute to dlang though. I suggested it years ago but there was no DIP process back then.

Reply:
https://forum.dlang.org/post/mailman.4463.1420983153.9932.digitalmars-d@puremagic.com